foyer/xml_writer.py

```diff
@@ -0,0 +1,281 @@
+            bond_force.set('id2', str(bond.atom2.idx))
+        for id in range(2):
+            bond_force.set('type{}'.format(id+1), atypes[id])
+        bond_force.set('length', str(round(bond.type.req / 10, 4)))
+        bond_force.set('k', str(round(bond.type.k * 4.184 * 200, 1)))
+
+def _write_angles(root, angles, unique):
+    angle_forces = ET.SubElement(root, 'HarmonicAngleForce')
+    for angle in angles:
+        angle_force = ET.SubElement(angle_forces, 'Angle')
+        atypes = [atype for atype in [angle.atom1.type,
+                                      angle.atom2.type,
+                                      angle.atom3.type]]
+        if unique:
+            # Sort the first and last atom types
+            atypes[::2] = sorted(atypes[::2])
+        else:
+            angle_force.set('id1', str(angle.atom1.idx))
+            angle_force.set('id2', str(angle.atom2.idx))
+            angle_force.set('id3', str(angle.atom3.idx))
+        for id in range(3):
+            angle_force.set('type{}'.format(id+1), atypes[id])
+        angle_force.set('angle', str(round(angle.type.theteq * (np.pi / 180), 10)))
+        angle_force.set('k', str(round(angle.type.k * 4.184 * 2, 3)))
+
+def _write_periodic_torsions(root, dihedrals, unique):
+    periodic_torsion_forces = ET.SubElement(root, 'PeriodicTorsionForce')
+    for dihedral in dihedrals:
+        if dihedral.improper:
+            dihedral_type = 'Improper'
+        else:
+            dihedral_type = 'Proper'
+        dihedral_force = ET.SubElement(periodic_torsion_forces, dihedral_type)
+        atypes = [atype for atype in [dihedral.atom1.type,
+                                      dihedral.atom2.type,
+                                      dihedral.atom3.type,
+                                      dihedral.atom4.type]]
+        if dihedral.improper:
+            # We want the central atom listed first and then sort the
+            # remaining atom types.
+            atypes[0], atypes[2] = atypes[2], atypes[0]
+            if unique:
+                atypes[1:] = sorted(atypes[1:])
+            else:
+                dihedral_force.set('id1', str(dihedral.atom3.idx))
+                dihedral_force.set('id2', str(dihedral.atom2.idx))
+                dihedral_force.set('id3', str(dihedral.atom1.idx))
+                dihedral_force.set('id4', str(dihedral.atom4.idx))
+        else:
+            if unique:
+                if atypes[0] > atypes[-1]:
+                    atypes = atypes[::-1]
+            else:
+                dihedral_force.set('id1', str(dihedral.atom1.idx))
+                dihedral_force.set('id2', str(dihedral.atom2.idx))
+                dihedral_force.set('id3', str(dihedral.atom3.idx))
+                dihedral_force.set('id4', str(dihedral.atom4.idx))
+        for id in range(4):
+            dihedral_force.set('type{}'.format(id+1), atypes[id])
+        dihedral_force.set('periodicity1', str(dihedral.type.per))
```

Comment:
Okay this is where things get hairy. In these periodic torsions (like rbtorsions), you can essentially have multiple terms (sum of a series). This chunk of code is only looking at one term within the series, when sometimes there are multiple (like line 96493 in  [here](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/ffxml/charmm36_nowaters.xml))

UNLIKE rbtorsions, this series does not get lumped into a single `structure.dihedral` term like it does in `structure.rb_torsions`. If there's a periodictorsion with multiple periodicities in the XML element, that gets "unwrapped" into multiple `structure.dihedral` terms (it looks like they're still adjacent elements in the `structure.dihedrals` list, though).

So I think there needs to be some sort of check for multiple periodicities by looking at adjacent terms. Sometimes there will be, sometimes there won't be. If there ARE multiple periodicities, then they should all get set in the same `periodictorsion` xml element.

The next problem (I need to investigate this further), is that the parameterized-structures all have `dihed.funct` == '1', which is okay for single-term periodictorsions, but should be `dihed.funct` == '9' for multi-term periodictorsions (I think, doing some comparisons with some of my charmm phospholipid parameters)

Reply:
That last problem with the `dihed.funct` might be a parmed issue, but i'll get back to this